browser/browsing_data/brave_clear_browsing_data.cc

```diff
@@ -0,0 +1,162 @@
+ public:
+  BrowsingDataRemovalWatcher() : observer_(this) {}
+
+  void ClearBrowsingDataForLoadedProfiles(
+      BraveClearBrowsingData::OnExitTestingCallback* testing_callback);
+
+  // BrowsingDataRemover::Observer implementation.
+  void OnBrowsingDataRemoverDone() override;
+
+ private:
+  bool GetClearBrowsingDataOnExitSettings(const Profile* profile,
+                                          int* remove_mask,
+                                          int* origin_mask);
+  void Wait();
+
+  int num_profiles_to_clear_ = 0;
+  base::RunLoop run_loop_;
+  ScopedObserver<content::BrowsingDataRemover,
+                 content::BrowsingDataRemover::Observer>
+      observer_;
+};
+
+bool BrowsingDataRemovalWatcher::GetClearBrowsingDataOnExitSettings(
+    const Profile* profile,
+    int* remove_mask,
+    int* origin_mask) {
+  DCHECK(remove_mask);
+  DCHECK(origin_mask);
+  const PrefService* prefs = profile->GetPrefs();
+  *remove_mask = 0;
+  *origin_mask = 0;
+
+  int site_data_mask = ChromeBrowsingDataRemoverDelegate::DATA_TYPE_SITE_DATA;
+  // Don't try to clear LSO data if it's not supported.
+  if (!prefs->GetBoolean(prefs::kClearPluginLSODataEnabled))
+    site_data_mask &= ~ChromeBrowsingDataRemoverDelegate::DATA_TYPE_PLUGIN_DATA;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteBrowsingHistoryOnExit) &&
+      prefs->GetBoolean(prefs::kAllowDeletingBrowserHistory))
+    *remove_mask |= ChromeBrowsingDataRemoverDelegate::DATA_TYPE_HISTORY;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteDownloadHistoryOnExit) &&
+      prefs->GetBoolean(prefs::kAllowDeletingBrowserHistory))
+    *remove_mask |= content::BrowsingDataRemover::DATA_TYPE_DOWNLOADS;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteCacheOnExit))
+    *remove_mask |= content::BrowsingDataRemover::DATA_TYPE_CACHE;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteCookiesOnExit)) {
+    *remove_mask |= site_data_mask;
+    *origin_mask |= content::BrowsingDataRemover::ORIGIN_TYPE_UNPROTECTED_WEB;
+  }
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeletePasswordsOnExit))
+    *remove_mask |= ChromeBrowsingDataRemoverDelegate::DATA_TYPE_PASSWORDS;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteFormDataOnExit))
+    *remove_mask |= ChromeBrowsingDataRemoverDelegate::DATA_TYPE_FORM_DATA;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteHostedAppsDataOnExit)) {
+    *remove_mask |= site_data_mask;
+    *origin_mask |= content::BrowsingDataRemover::ORIGIN_TYPE_PROTECTED_WEB;
+  }
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteMediaLicensesOnExit))
+    *remove_mask |= content::BrowsingDataRemover::DATA_TYPE_MEDIA_LICENSES;
+
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteSiteSettingsOnExit))
+    *remove_mask |=
+        ChromeBrowsingDataRemoverDelegate::DATA_TYPE_CONTENT_SETTINGS;
+
+  return (*remove_mask != 0);
+}
+
+void BrowsingDataRemovalWatcher::ClearBrowsingDataForLoadedProfiles(
```

Comment:
Might be good to include very brief comments explaining what this does (either here or above the class definition)
ex: `Create tasks for clearing each applicable type of data (user specified) and block until all finish. When each type of clear finishes, it decrements a counter and is "done" when it reaches 0`